content/posts/include-vs-inherit.md

```diff
@@ -0,0 +1,231 @@
+# call below. It does the following:
+#  1. Inherits all of the variables from $1.
+#  2. Records the inheritance in the .INHERITS_FROM variable
+#
+# (2) and the PRODUCTS variable can be used together to reconstruct the include hierarchy
+# See e.g. product-graph.mk for an example of this.
+#
+define inherit-product
+  $(eval _inherit_product_wildcard := $(wildcard $(1)))\
+  $(if $(_inherit_product_wildcard),,$(error $(1) does not exist.))\
+  $(foreach part,$(_inherit_product_wildcard),\
+    $(if $(findstring ../,$(part)),\
+      $(eval np := $(call normalize-paths,$(part))),\
+      $(eval np := $(strip $(part))))\
+    $(foreach v,$(_product_var_list), \
+        $(eval $(v) := $($(v)) $(INHERIT_TAG)$(np))) \
+    $(eval current_mk := $(strip $(word 1,$(_include_stack)))) \
+    $(eval inherit_var := PRODUCTS.$(current_mk).INHERITS_FROM) \
+    $(eval $(inherit_var) := $(sort $($(inherit_var)) $(np))) \
+    $(call dump-inherit,$(strip $(word 1,$(_include_stack))),$(1)) \
+    $(call dump-config-vals,$(current_mk),inherit))
+endef
+```
+
+The comment above the implementation should help, but actually it creates more confusion if we compare it with the actual implementation. We can see that the code is checking if the file exists, if does some path normalization if needed and then it actually looks for all the variables into a product_var_list. Hmmmm...
+
+So it does not inherit **all** the variables, just the ones in that list. It seems to be defined in the same product.mk file as the function:
+```
+product_var_list :=$= $(_product_single_value_vars) $(_product_list_vars)
+```
+So basically we have 2 lists, one with single value variables and one with list variables, and the common part if that all of them starts with **PRODUCT_**. I will not list them here because they might change, I recommend to have a look into the AOSP file to check them out. There are 4 exceptions to this naming rule:
+```
+DEVICE_PACKAGE_OVERLAYS
+VENDOR_PRODUCT_RESTRICT_VENDOR_FILES
+VENDOR_EXCEPTION_MODULES
+VENDOR_EXCEPTION_PATHS
+```
+
+I found one interesting comment in that file that is relevant for the single value variables, and how their inheritance will work (will put this to the test later):
+
+```
+# Variables that are meant to hold only a single value.
+# - The value set in the current makefile takes precedence over inherited values
+# - If multiple inherited makefiles set the var, the first-inherited value wins
+```
+
+Seems pretty straightforward now, but let's put this to the test. I created a small test device that is inheriting from the AOSP Car Emulator device, you can get all the needed files from this [archive](device.esrlabs.tar.gz) which you should unpack in your device/ directory.
+
+The main part is in esr_emulator.mk, where we try to inherit some other files:
+
+```
+$(call inherit-product, device/generic/car/emulator/aosp_car_emulator.mk)
+$(call inherit-product, $(SRC_TARGET_DIR)/product/aosp_x86_64.mk)
```

Comment:
Could we mention what SRC_TARGET_DIR corresponds to? Actually I don't know myself right now ;-)